### ISSUE TITLE
Option for disabling request canceling if Non-IOException occurred in interceptor

### ISSUE THREAD

**Issue by ntngel1** [square/okhttp#5731]
Based on this [issue](https://github.com/gildor/kotlin-coroutines-retrofit/issues/34). If we're throwing exceptions inside interceptor, they should inherit IOException to be properly handled without crashing the app. But there's some situations in which you don't want to inherit IOException.
One of this cases: we are handling errors of API. Not HTTP errors. API returns some templated entities with status-code which can be checked to detect some errors like invalid parameters that was passed. 
And one of error handling solutions is to create interceptor that checks all of requests' responses and throwing domain exceptions if status code indicates error.
I suggest to add some parameter to OkHttpClient.Builder that will allow us to disable canceling requests if some Non-IOException was thrown inside interceptor.

**Comment by swankjesse** [comment]
What would it do instead? Somebody needs to close the response and release resources.

**Comment by ntngel1** [comment]
Sorry, I put it wrong. My idea is not to crash the app when Non-IOException occurred inside interceptor.

**Comment by yschimke** [comment]
Isn't this fixed in 4.3 (and recent 3.x releases) via clean cancellation on a non-IOException?

https://github.com/square/okhttp/issues/5151

**Comment by swankjesse** [comment]
I think this is a request for us to not notify the uncaught exception handler.

@ntngel1 can you install your own uncaught exception handler?

**Comment by yschimke** [comment]
Or catch inside your own interceptor. I think it's tough for us to do this without swallowing real problems like logic bugs or version incompatibilities and silently retrying.

**Comment by ntngel1** [comment]
Again, I'll try to explain what do I want.
I have OkHttpClient with some interceptor that is processing response from server. If this response doesn't match some requirements, it throws non-IOException, and I expect it to be possible to catch it inside my repository, but it occurrs app crash. I don't understand how can I achieve this behaviour with UncaughtExceptionHandler.

There were no problems with throwing non-IOExceptions when I was using RxJava with Retrofit, but when I got in coroutines with retrofit, I detected this problem.
